### ISSUE TITLE
Idea: use a genuine Linux PTY via WSL on Windows

### ISSUE THREAD

**Issue by chris-morgan** [alacritty/alacritty#3707]
WinPTY and ConPTY keep on causing me misery, mangling perfectly good ANSI escape codes and slowing things down for literally no benefit for me, given that I’m running Linuxy stuff only in my terminal now.

So I’m finally wondering: would it be possible to provide another backend for Windows (besides the WinPTY and ConPTY ones), which would dive into WSL and spawn a PTY the Linux way, and use that as its source?

I have no idea if this is feasible or how it would be done. If it worked, I have no idea whether you could still spawn Windows executables from the Linux side, or whether they would mysteriously stop working. (I’d be *totally fine* with that, but it would put a damper on things for general utility.)

If given somewhere to start, I’d be interested in working on this.

Objectives of this would be: *vastly* improved performance (ConPTY is faster than WinPTY was, but it’s still a major bottleneck), and pristine ANSI codes so I can be free of problems due to the likes of #3361, https://github.com/microsoft/terminal/issues/2661 and https://github.com/microsoft/terminal/issues/5682. And I could finally switch to a light terminal theme without major problems, I think.

**Comment by kchibisov** [comment]
>So I’m finally wondering: would it be possible to provide another backend for Windows (besides the WinPTY and ConPTY ones), which would dive into WSL and spawn a PTY the Linux way, and use that as its source?

I have no idea how to do that (as well as other things on Windows), however we can accept more backends for PTY reading.

>If given somewhere to start, I’d be interested in working on this.

maybe some other Wnidows folks know something about it. Though new pty backend should be in https://github.com/alacritty/alacritty/tree/master/alacritty_terminal/src/tty .

cc @davidhewitt ?

**Comment by davidhewitt** [comment]
Hmm, interesting. I don't know whether it's possible to access any WSL in/out handles directly from a Win32 program. Probably `conhost.exe` would always get in the way to try and host the WSL process?

**Comment by chris-morgan** [comment]
One approach that might work: run a helper ELF binary which establishes the actual PTY, and establish a pipe between the two. I’m not sure how that’s done, but I know things like weasel-pageant works that way, so I could look into how they do it.

There’s a moderate ergonomics cost to this (“compile this Linux helper program and drop it beside alacritty.exe”), but it should make it possible.

I guess this could actually scale to a full remote PTY system—attach to an existing PTY or some such thing. Not sure such a thing would be *useful* for all that much, but it’s a beautiful theoretical construct.

**Comment by davidhewitt** [comment]
> One approach that might work: run a helper ELF binary which establishes the actual PTY, and establish a pipe between the two. I’m not sure how that’s done, but I know things like weasel-pageant works that way, so I could look into how they do it.

I once suggested an idea similar to this of bundling [a precompiled version of `OpenConsole.exe`](https://github.com/alacritty/alacritty/issues/3085#issuecomment-564407410) directly alongside alacritty. IMO if we're having frustration with ConPTY (and particularly because we currently rely on Windows' bundled version rather than being able to pin to a best compatible version), we should solve it that way rather than bundling an ELF binary. This would support the whole windows ecosystem rather than just WSL, and if we're having any issues with ConPTY we could fix them upstream rather than maintain extra code here.

I'm holding out for the pipe dream that one day ConPTY will also support "passthrough mode", which should allow everything which doesn't need the legacy windows console APIs to be able to just write VT which ConPTY will simply forward to Alacritty. https://github.com/microsoft/terminal/issues/1173

**Comment by chrisduerr** [comment]
Seems like this issue is just going to be resolved once WSL2 lands eventually?

**Comment by davidhewitt** [comment]
> Seems like this issue is just going to be resolved once WSL2 lands eventually?

How so?

**Comment by chrisduerr** [comment]
Since WSL2 is a proper VM, that should easily allow launching an Alacritty instance that hooks up directly to a Linux PTY, right?

**Comment by davidhewitt** [comment]
Oh quite possibly! I'm not sure how easy it is to get an X window server from WSL 2 though :)

**Comment by chrisduerr** [comment]
I thought getting an X server from WSL2 wasn't an issue.

**Comment by chris-morgan** [comment]
Allowing launching an Alacritty instance that hooks up directly to a Linux PTY is exactly what I’m suggesting. But I don’t know why you’re bringing X into it (the typical approach for which is to run the X server on Windows as VcXsrv, and just set `DISPLAY=:0` on the Linux side, and it just works), because the Alacritty GUI needs to be on the Windows side to work properly, e.g. to support a mixed-DPI environment.

**Comment by chrisduerr** [comment]
I mean there is no Linux PTY on Windows natively. So you can't just attach to a Linux PTY on Windows.

Everything else you could do would just be a giant mess and adding another PTY backend to Windows definitely isn't an option. Since at that point we'd just have three broken backends.

So as far as I can see, the best option would be to just run Alacritty on Linux within Windows if you're looking for a proper Linux PTY. If you want a Linux terminal working just like it does on Linux, maybe run Linux?

**Comment by jerch** [comment]
Imho all it needs is forwarding a named pipe from windows into WSL and attach it there to the pty master end (moving byte chunks from one side to the other). Plus forwarding a few ioctls like resize handling.

This might be a starting point for the pipe forwarding https://github.com/jstarks/npiperelay (prolly not working with WSL2, idk).

**Comment by chrisduerr** [comment]
That sounds like something I definitely would **not** want to have anywhere near Alacritty.

**Comment by chris-morgan** [comment]
Could you explain why not? Because I don’t see anything wrong with the approach.

**Comment by chrisduerr** [comment]
Because it's a fragile approach that is in no way properly supported by Microsoft and requires a completely separate path for Alacritty to the existing PTYs. Two broken Windows PTYs with half a million bugs for each that we cannot resolve is plenty. I have zero interest in a third.

**Comment by chris-morgan** [comment]
It doesn’t sound fragile to me, if it’s as simple as forwarding one socket across platform boundaries and figuring out a way of spawning the PTY. That sounds to have a *vastly* smaller surface area than ConPTY, and it’d allow Alacritty to work the same way on Windows (well, modulo the “it’s actually Linux underneath” bit) as it does on other platforms.

**Comment by jerch** [comment]
@chris-morgan I did some early tests with `openpty` in wsl python to see, whether WSL creates a fully functional pty interface and yepp - this seems to be the case. Also `tmux` and `screen` in WSL run quite smooth with correct Unicode and colors annotated.

But the devil is in the details, as always. WSL allows native windows programs to be called in the WSL context, and they even spit out content to a WSL terminal, if they are meant as console apps. To achieve that, windows (ConPty again?) shims in and does some native console API scraping, and forwards that to the WSL terminal. I think this is also done by ConPty, because it shows all the nasty "fullscreen repaint" and cursor quirks we see on ConPty in general. Which makes native windows console apps second class citizens on a WSL driven PTY. Which is not a big bummer in my opinion, as windows console apps are not that great after 20ys of strict GUI promotion. Still MS did the opposite with ConPty mangling even real WSL pty tools through that weird abstraction.

So yes windows has with WSL first class pty support, incl. programs, but only for WSL programs.

Now what to make out of it? I am for myself quite frustrated with the current ConPty state, we get constantly issue reports about it for broken Unicode/emojis, limited colors, weird screen updates and what else. Really annoying. Still have to do a few more investigations, yet bridging a real PTY fd through to windows seems quite a promising from my perspective atm. It would give all WSL apps full pty support as under any POSIX system while the emulator can run in windows realms (better perf), and well windows native console apps gonna suck as they did before. Whether this can be done reliably is not yet clear to me, I might try to do a WSL-bridged mode in `node-pty`.

**Comment by davidhewitt** [comment]
For those wanting to run `alacritty` from WSL, looks like this might be a supported feature of windows later this year: https://www.theverge.com/2020/5/19/21263377/microsoft-windows-10-linux-gui-apps-gpu-acceleration-wsl-features

This seems like a much nicer option than trying to do some hybrid where the alacritty GUI runs as a windows app connected to a WSL process.

**Comment by Walther** [comment]
Here's a funny / silly / slightly cursed workaround: when you launch `alacritty` and you get the PowerShell, you can actually run `bash.exe` in it to drop into your WSL.
![image](https://user-images.githubusercontent.com/2943750/85169413-6e051a00-b274-11ea-8c52-592b78d6c58d.png)

**Comment by chris-morgan** [comment]
@Walther That doesn’t solve the problems at all: ConPTY or WinPTY are still in the way, mangling everything.

**Comment by chrisduerr** [comment]
Yeah, this is about the *PTY*, not the shell. Being able to run different shells from different environments has always worked and that's intentional not a 'cursed workaround'. That's just how the PTY/Shell separation works.

**Comment by Walther** [comment]
Sorry for my misunderstanding! Thanks for the clarification 🙇‍♂️

**Comment by tactlessfish** [comment]
I [read about](https://github.com/mintty/wsltty/issues/97) a project called [wslbridge](rprichard/wslbridge) that could be useful for this.

> wsltty uses wslbridge which is a dedicated variation of winpty with benefits in the handling of escape sequences, see rprichard/winpty#89 for the motivation to create wslbridge.

**Comment by sebastianstudniczek** [comment]
Recently I stumpled across such ghostty based projects which are focusing on being WSL native so they they have option to bypass ConPTY.

- https://github.com/sudo-tee/hollow/blob/main/docs/platforms/wsl.md

> The way I did it in my project was simply to open wsl.exe as a normal executable and spawn a simple binary to stream pty bytes directly with a framed protocol over OSC.

- https://github.com/Codavo/ghostinthewsl
> What I did instead was create a little portable bridge binary which uses Hyper-V/VSOCK to communicate. I think that's cleaner, but it's more complicated.

Discussion: https://github.com/ghostty-org/ghostty/discussions/12371#discussioncomment-16751077
